### PR TITLE
Asynchronous load of emoji list

### LINKED ISSUE #2036: Provide asynchronous emoji list loading
## Type of report

Task

## Provide description of the task

Provide possibility to load asynchronously emoji list for emoji plugin.

## Other details

* Browser: …
* OS: …
* CKEditor version: …
* Installed CKEditor plugins: …

### PR DESCRIPTION
**Pull request by msamsel**
## What is the purpose of this pull request?

New feature

## Does your PR contain necessary tests?

All patches which change the editor code must include tests. You can always read more
on [PR testing](https://docs.ckeditor.com/ckeditor4/docs/#!/guide/dev_contributing_code-section-tests),
[how to set the testing environment](https://docs.ckeditor.com/ckeditor4/docs/#!/guide/dev_tests) and
[how to create tests](https://docs.ckeditor.com/ckeditor4/docs/#!/guide/dev_tests-section-creating-your-own-test)
in the official CKEditor documentation.

### This PR contains

- [x] Unit tests
- [x] Manual tests

## What changes did you make?

- provide asynchronous load of emojis
- provide configuration option to load custom emoji list

Close #2036

### CODE REVIEW

File: plugins/emoji/plugin.js [hunk 1]
```diff
@@ -158,3 +174,34 @@
  * @cfg {Number} [emoji_minChars = 2]
  * @member CKEDITOR.config
  */
+
+/**
+ * Address to json file containing emoji list. File is downloaded through {@link CKEDITOR.ajax#load} method
```
Comment (mlewand) [inline plugins/emoji/plugin.js]: JSON is an acronym should be uppercased.

**Comment by msamsel** [comment]
I've rebase branch to contain change introduced in #2032

**Comment by mlewand** [comment]
Issue reported by @Comandeer is valid. What's more imagine a case when CKEditor makes few snapshots before the `:emoji_tokens:` are replaced. That would mean that we'd have to change an entire snapshots history.

Instead let's make a tough decision and drop the feature which replaced valid emoji tokens in the content, at paste, drop, etc.

Ideally I'd like to keep this feature at `paste` event, however looking at the implementation, it's implemented using core text filter mechanics and it would require some refactoring to make it happen. We want to conclude it so **instead we'll remove the feature of replacing text tokens** with emoji entirely.

To sum it up:

* Keep the async emoji loading
* Remove parts of emoji that are replacing text tokens with real emojis on setData/insertText/paste/drop etc.

File: plugins/emoji/plugin.js [hunk 2]
```diff
@@ -11,132 +11,96 @@
...
 
-			function matchCallback( text, offset ) {
-				var left = text.slice( 0, offset ),
-					match = left.match( new RegExp( ':\\S{' + charactersToStart + '}\\S*$' ) );
+				var emojiList = editor._.emoji.list,
+					charactersToStart = editor.config.emoji_minChars === undefined ? 2 : editor.config.emoji_minChars;
 
-				if ( !match ) {
-					return null;
+				if ( editor.status !== 'ready' ) {
```
Comment (Comandeer) [inline plugins/emoji/plugin.js]: Is this `if` needed now, when substitution is removed and initialisation just creates autocomplete instance?

File: plugins/emoji/plugin.js [hunk 3]
```diff
@@ -11,132 +11,96 @@
...
 
-			function isEmojiAllowed( range ) {
-				var elementsPath,
-					editable = editor.editable();
-				if ( range ) {
-					elementsPath = new CKEDITOR.dom.elementPath( range.startContainer, editable );
-					return elementsPath.contains( forbiddenScope ) ? false : true;
-				} else {
-					return true;
+				function initPlugin() {
```
Comment (Comandeer) [inline plugins/emoji/plugin.js]: If we remove the `if` mentioned above, then this function could be inlined.

File: tests/plugins/emoji/customemojilist.js [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+	bender.test( {
+		setUp: function() {
+			if ( CKEDITOR.env.ie && CKEDITOR.env.version < 9 ) {
+				assert.ignore();
+			}
+		},
+
+		// #2036
+		'test custom emoji list is loadd': function() {
```
Comment (Comandeer) [inline tests/plugins/emoji/customemojilist.js]: Typo: loadd → load

File: tests/plugins/emoji/customemojilist.js [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+			emojiTools.runAfterInstanceReady( editor, null, function( editor ) {
+				assert.areSame( 1, editor._.emoji.list.length );
+				objectAssert.areEqual( { id: ':star:', symbol: '⭐' }, editor._.emoji.list[ 0 ] );
+			} );
+		},
+
+		'test invalid emoji list': function() {
+			var editor = this.editors.classic2;
+			emojiTools.runAfterInstanceReady( editor, null, function( editor ) {
+				assert.isUndefined( editor._.emoji, 'There are created emoji private data, so emoji was loaded what is wrong for this case.' );
```
Comment (Comandeer) [inline tests/plugins/emoji/customemojilist.js]: This doesn't seem like good assertion message.

File: plugins/emoji/plugin.js (hunk 2 above)
Comment (msamsel) [reply plugins/emoji/plugin.js]: yes, it's required because of this bug: https://github.com/ckeditor/ckeditor-dev/issues/2114

Inlining this function cause random error, where plugin might be not initialised during loading stage (depend how fast json will be loaded). Error was clearly visible in Firefox browser.

File: plugins/emoji/plugin.js (hunk 3 above)
Comment (msamsel) [reply plugins/emoji/plugin.js]: related to previous comment :)

**Comment by msamsel** [comment]
Font injection for IE browser in divarea editor is extracted here: https://github.com/ckeditor/ckeditor-dev/issues/2121

**Comment by msamsel** [comment]
I correct those tests. :)
